clang/test/CodeGenSYCL/add_ir_annotations_member_pointer_access.cpp

```diff
@@ -0,0 +1,63 @@
+// RUN: %clang_cc1 -internal-isystem %S/Inputs -disable-llvm-passes \
+// RUN:    -triple spir64-unknown-unknown -fsycl-is-device -S \
+// RUN:    -opaque-pointers -emit-llvm %s -o - | FileCheck %s
+// RUN: %clang_cc1 -internal-isystem %S/Inputs -disable-llvm-passes \
+// RUN:    -triple spir64-unknown-unknown -fsycl-is-device -DTEST_SCALAR -S \
+// RUN:    -opaque-pointers -emit-llvm %s -o - | FileCheck %s
+
+// Tests the generation of IR annotation calls from
+// __sycl_detail__::add_ir_annotations_member_pointer_access attributes.
+// This gurantee that the ptr.annotation is inserted before
+// the ptr-to-ptr load
+
+#include "mock_properties.hpp"
+#include "sycl.hpp"
+
+#define TEST_T int*
+
+template <typename... Properties> class g {
+public:
+  TEST_T x
+#ifdef __SYCL_DEVICE_ONLY__
+      [[__sycl_detail__::add_ir_annotations_member(Properties::name..., Properties::value...)]]
+#endif
+      ;
+
+  g() : x() {}
+  g(TEST_T _x) : x(_x) {}
+};
+
+int main() {
+  sycl::queue q;
+  g<prop1, prop2, prop3, prop4, prop5, prop6, prop7, prop8> a;
+  q.submit([&](sycl::handler &h) {
+    h.single_task<class test_kernel1>(
+        [=]() {
+          *a.x = 1;
+        });
+  });
+  return 0;
+}
+
+// CHECK-DAG: @[[AnnotName:.*]] = private unnamed_addr addrspace(1) constant [16 x i8] c"sycl-properties\00", section "llvm.metadata"
```

Comment:
There is no change of the IR before and after my PR. This is a new test added to guarantee that the form of IR that is generated will not break

Reply:
Can you clarify what exactly the pattern you are searching for is? I would also suggest using CHECK/CHECK-NEXT if you care about code pattern so its less fragile.

Reply:
I would suggest simplifying this test to make it easier to read what exactly you are testing for. You just need 1 property to test this right?

I agree you don't need to check CHECK-NEXT since that will result in a lot more check lines. But CHECK (without the next) will ensure order or pattern you are testing making this test more robust in my opinion. 

I'm also not sure if you need to add a non-opaque pointer test for this. @premanandrao what is the protocol now? Do we add an opaque pointer and a non-opaque pointer test for all CodeGen tests?

Reply:
+1 for using CHECK.

@elizabethandrews, unless I am mistaken, we have not yet switched over to opaque-pointers in all phases yet.   So I think, we should continue adding both versions for CodeGen tests.

Reply:
Do you need to test with 8 properties?

Reply:
This is copied from another file add_ir_annotation_member.cpp

Reply:
Right, but I assume that was testing properties being applied in different ways and this test isn't really about that and so IMO doesn't need all of this. Anyway its not a big deal. I just tend to keep tests as small as possible to easily see what is being tested